src/Nethermind/Nethermind.State.Test/StorageProviderTests.cs

```diff
@@ -236,6 +236,171 @@ public void Can_commit_when_exactly_at_capacity_regression()
+            provider.Set(new StorageCell(ctx.Address1, 1), _values[9]);
+            snapshots[2] = ((IStorageProvider)provider).TakeSnapshot();
+            Assert.AreEqual(snapshots[2].TransientStorageSnapshot, 1);
+            Assert.AreEqual(snapshots[2].PersistentStorageSnapshot, 0);
+
+            // Only update persistent
+            provider.Set(new StorageCell(ctx.Address1, 1), _values[8]);
+            snapshots[3] = ((IStorageProvider)provider).TakeSnapshot();
+            Assert.AreEqual(snapshots[3].TransientStorageSnapshot, 1);
+            Assert.AreEqual(snapshots[3].PersistentStorageSnapshot, 1);
+
+            provider.Restore(snapshots[snapshot + 1]);
+
+            // Since we didn't update transient on the 3rd snapshot
+            if (snapshot == 2)
+            {
+                snapshot--;
+            }
+            Assert.AreEqual(_values[snapshot + 1], provider.GetTransientState(new StorageCell(ctx.Address1, 1)));
+        }
+
+        [TestCase(-1)]
+        [TestCase(0)]
+        [TestCase(1)]
+        [TestCase(2)]
+        public void Permanent_state_restores_independent_of_transient_state(int snapshot)
+        {
+            Context ctx = new();
+            StorageProvider provider = BuildStorageProvider(ctx);
+            Snapshot[] snapshots = new Snapshot[4];
+
+            // No updates
+            snapshots[0] = ((IStorageProvider)provider).TakeSnapshot();
+            Assert.AreEqual(snapshots[0].TransientStorageSnapshot, -1);
+            Assert.AreEqual(snapshots[0].PersistentStorageSnapshot, -1);
+
+            // Only update persistent
+            provider.Set(new StorageCell(ctx.Address1, 1), _values[1]);
+            snapshots[1] = ((IStorageProvider)provider).TakeSnapshot();
+            Assert.AreEqual(snapshots[1].PersistentStorageSnapshot, 0);
+            Assert.AreEqual(snapshots[1].TransientStorageSnapshot, -1);
+
+            // Update both
+            provider.Set(new StorageCell(ctx.Address1, 1), _values[2]);
+            provider.SetTransientState(new StorageCell(ctx.Address1, 1), _values[9]);
+            snapshots[2] = ((IStorageProvider)provider).TakeSnapshot();
+            Assert.AreEqual(snapshots[2].PersistentStorageSnapshot, 1);
+            Assert.AreEqual(snapshots[2].TransientStorageSnapshot, 0);
+
+            // Only update transient
+            provider.SetTransientState(new StorageCell(ctx.Address1, 1), _values[8]);
+            snapshots[3] = ((IStorageProvider)provider).TakeSnapshot();
+            Assert.AreEqual(snapshots[3].PersistentStorageSnapshot, 1);
+            Assert.AreEqual(snapshots[3].TransientStorageSnapshot, 1);
+
+            provider.Restore(snapshots[snapshot + 1]);
+
+            // Since we didn't update persistent on the 3rd snapshot
+            if (snapshot == 2)
+            {
+                snapshot--;
+            }
+            Assert.AreEqual(_values[snapshot + 1], provider.Get(new StorageCell(ctx.Address1, 1)));
+        }
```

Comment:
Can those potentially be rewritten to subsequent TestCases that will operate on changing state?

Reply:
Not really I was thinking of rewriting the test to have one assert per test.